markdowns/19_Text_to_Image_search_pipeline_with_MultiModal_Retriever.md

```diff
@@ -0,0 +1,190 @@
+
+
+```python
+from haystack.document_stores import InMemoryDocumentStore
+
+# Here we initialize the DocumentStore with 512 dims 
+# as we will use it to store 512 dim image embeddings 
+document_store = InMemoryDocumentStore(embedding_dim=512)
+```
+
+# Downloading Data
+
+Download 18 sample images of different animals from . You can find them in data/tutorial19/spirit-animals/ as a set of .jpg files.
+
+
+```python
+from haystack.utils import fetch_archive_from_http
+
+doc_dir = "data/tutorial19"
+
+fetch_archive_from_http(
+    url="https://s3.eu-central-1.amazonaws.com/deepset.ai-farm-qa/datasets/documents/spirit-animals.zip",
+    output_dir=doc_dir
+)
+```
+
+Add the images you just downloaded into Haystack Document objects and write them into the DocumentStore.
+
+
+```python
+import os
+
+from haystack import Document
+
+images = [
+    Document(content=f"./{doc_dir}/spirit-animals/{filename}", content_type="image")
+    for filename in os.listdir(f"./{doc_dir}/spirit-animals/")
+]
+
+document_store.write_documents(images)
+```
+
+# Initializing the Retriever
+
+Retrievers sift through all the images and return only those that are relevant based on the input query. Here we are using the [OpenAI CLIP model](https://github.com/openai/CLIP/blob/main/model-card.md) to embed images. For more details on supported modalities, see [MultiModalRetriever](https://docs.haystack.deepset.ai/docs/retriever#multimodal-retrieval).
+
+
+```python
+from haystack.nodes.retriever.multimodal import MultiModalRetriever
+
+retriever_text_to_image = MultiModalRetriever(
+    document_store=document_store,
+    query_embedding_model = "sentence-transformers/clip-ViT-B-32",
+    query_type="text",
+    document_embedding_models = {"image": "sentence-transformers/clip-ViT-B-32"}
+)
+
+document_store.update_embeddings(retriever=retriever_text_to_image)
+```
+
+# Creating the MultiModal search Pipeline
+
+We are populating a Pipeline with a MultiModalRetriever node. This search pipeline queries the image database with text and returns the most relevant images.
+
+
+```python
+from haystack import Pipeline
+
+pipeline = Pipeline()
+pipeline.add_node(
+    component=retriever_text_to_image, 
+    name="retriever_text_to_image", 
+    inputs=["Query"]
+)
+```
+
+# Searching Through the Images
+
+Use the pipeline `run()` method to query the images in the DocumentStore. The query argument is where you type your text query. Additionally, you can set the number of images you want the MultiModalRetriever to return using the `top-k` parameter. To learn more about setting arguments, see [Pipeline Arguments](https://docs.haystack.deepset.ai/docs/pipelines#arguments).
+
+
+```python
+results = pipeline.run(
+    query="Animal who lives in the water",
+    params={"retriever_text_to_image": {"top_k": 3}}
+)
+
+results = sorted(results["documents"], key=lambda d: d.score, reverse=True)
```

Comment:
can we add a comment explaining what it does?

Reply:
Added. Although I think it's pretty standard python sorting which any Python user should know.